barretenberg/cpp/src/barretenberg/commitment_schemes/ipa/ipa.hpp

```diff
@@ -464,6 +464,122 @@ template <typename Curve_> class IPA {
+            transcript->template receive_from_prover<typename Curve::BaseField>("IPA:poly_degree_plus_1")
+                .get_value()); // note this is base field because this is a uint32_t, which should map
+                               // to a bb::fr, not a grumpkin::fr, which is a BaseField element for
+                               // Grumpkin
+
+        // Step 2.
+        // Receive generator challenge u and compute auxiliary generator
+        const Fr generator_challenge = transcript->template get_challenge<Fr>("IPA:generator_challenge");
+        auto builder = generator_challenge.get_context();
+
+        Commitment aux_generator = Commitment::one(builder) * generator_challenge;
+
+        const auto log_poly_degree = numeric::get_msb(static_cast<uint32_t>(poly_length));
+        // Step 3.
+        // Compute C' = C + f(\beta) ⋅ U
+        GroupElement C_prime = opening_claim.commitment + aux_generator * opening_claim.opening_pair.evaluation;
+
+        auto pippenger_size = 2 * log_poly_degree;
+        std::vector<Fr> round_challenges(log_poly_degree);
+        std::vector<Fr> round_challenges_inv(log_poly_degree);
+        std::vector<Commitment> msm_elements(pippenger_size);
+        std::vector<Fr> msm_scalars(pippenger_size);
+
+        // Step 4.
+        // Receive all L_i and R_i and prepare for MSM
+        for (size_t i = 0; i < log_poly_degree; i++) {
+            std::string index = std::to_string(log_poly_degree - i - 1);
+            auto element_L = transcript->template receive_from_prover<Commitment>("IPA:L_" + index);
+            auto element_R = transcript->template receive_from_prover<Commitment>("IPA:R_" + index);
+            round_challenges[i] = transcript->template get_challenge<Fr>("IPA:round_challenge_" + index);
+
+            round_challenges_inv[i] = round_challenges[i].invert();
+
+            msm_elements[2 * i] = element_L;
+            msm_elements[2 * i + 1] = element_R;
+            msm_scalars[2 * i] = round_challenges_inv[i];
+            msm_scalars[2 * i + 1] = round_challenges[i];
+        }
+
+        // Step 5.
+        // Compute C₀ = C' + ∑_{j ∈ [k]} u_j^{-1}L_j + ∑_{j ∈ [k]} u_jR_j
+        GroupElement LR_sums = GroupElement::batch_mul(msm_elements, msm_scalars);
+
+        GroupElement C_zero = C_prime + LR_sums;
+
+        //  Step 6.
+        // Compute b_zero where b_zero can be computed using the polynomial:
+        //  g(X) = ∏_{i ∈ [k]} (1 + u_{i-1}^{-1}.X^{2^{i-1}}).
+        //  b_zero = g(evaluation) = ∏_{i ∈ [k]} (1 + u_{i-1}^{-1}. (evaluation)^{2^{i-1}})
+
+        Fr b_zero = Fr(1);
+        for (size_t i = 0; i < log_poly_degree; i++) {
+            b_zero *= Fr(1) + (round_challenges_inv[log_poly_degree - 1 - i] *
+                               opening_claim.opening_pair.challenge.pow(1 << i));
+        }
+
+        // Step 7.
+        // Construct vector s
+        std::vector<Fr> s_vec(poly_length);
+
+        for (size_t i = 0; i < poly_length; i++) {
```

Comment:
Add a TODO to do this in a tree

Reply:
There already was one in the parallelised function, copied here as well:)